subsys/bluetooth/mesh/crypto_mtls.c

```diff
@@ -0,0 +1,644 @@
+	0xee, 0x7f, 0x0d, 0x9c, 0xe9, 0x78, 0x0a, 0x9b,
+	0xfc, 0x6d, 0x1f, 0x8e, 0xfb, 0x6a, 0x18, 0x89,
+	0xf2, 0x63, 0x11, 0x80, 0xf5, 0x64, 0x16, 0x87,
+
+	0xd8, 0x49, 0x3b, 0xaa, 0xdf, 0x4e, 0x3c, 0xad,
+	0xd6, 0x47, 0x35, 0xa4, 0xd1, 0x40, 0x32, 0xa3,
+	0xc4, 0x55, 0x27, 0xb6, 0xc3, 0x52, 0x20, 0xb1,
+	0xca, 0x5b, 0x29, 0xb8, 0xcd, 0x5c, 0x2e, 0xbf,
+
+	0x90, 0x01, 0x73, 0xe2, 0x97, 0x06, 0x74, 0xe5,
+	0x9e, 0x0f, 0x7d, 0xec, 0x99, 0x08, 0x7a, 0xeb,
+	0x8c, 0x1d, 0x6f, 0xfe, 0x8b, 0x1a, 0x68, 0xf9,
+	0x82, 0x13, 0x61, 0xf0, 0x85, 0x14, 0x66, 0xf7,
+
+	0xa8, 0x39, 0x4b, 0xda, 0xaf, 0x3e, 0x4c, 0xdd,
+	0xa6, 0x37, 0x45, 0xd4, 0xa1, 0x30, 0x42, 0xd3,
+	0xb4, 0x25, 0x57, 0xc6, 0xb3, 0x22, 0x50, 0xc1,
+	0xba, 0x2b, 0x59, 0xc8, 0xbd, 0x2c, 0x5e, 0xcf
+};
+
+uint8_t bt_mesh_fcs_calc(const uint8_t *data, uint8_t data_len)
+{
+	uint8_t fcs = 0xff;
+
+	while (data_len--) {
+		fcs = crc_table[fcs ^ *data++];
+	}
+
+	BT_DBG("fcs 0x%02x", 0xff - fcs);
+
+	return 0xff - fcs;
+}
+
+bool bt_mesh_fcs_check(struct net_buf_simple *buf, uint8_t received_fcs)
+{
+	const uint8_t *data = buf->data;
+	uint16_t data_len = buf->len;
+	uint8_t fcs = 0xff;
+
+	while (data_len--) {
+		fcs = crc_table[fcs ^ *data++];
+	}
+
+	return crc_table[fcs ^ received_fcs] == 0xcf;
+}
+
+int bt_mesh_virtual_addr(const uint8_t virtual_label[16], uint16_t *addr)
+{
+	uint8_t salt[16];
+	uint8_t tmp[16];
+	int err;
+
+	err = bt_mesh_s1("vtad", salt);
+	if (err) {
+		return err;
+	}
+
+	err = bt_mesh_aes_cmac_one(salt, virtual_label, 16, tmp);
+	if (err) {
+		return err;
+	}
+
+	*addr = (sys_get_be16(&tmp[14]) & 0x3fff) | 0x8000;
+
+	return 0;
+}
+
+int bt_mesh_prov_salt(const uint8_t conf_salt[16], const uint8_t prov_rand[16],
+		      const uint8_t dev_rand[16], uint8_t prov_salt[16])
+{
+	const uint8_t prov_salt_key[16] = { 0 };
+	struct bt_mesh_sg sg[] = {
+		{ conf_salt, 16 },
+		{ prov_rand, 16 },
+		{ dev_rand, 16 },
+	};
+
+	return bt_mesh_aes_cmac(prov_salt_key, sg, ARRAY_SIZE(sg), prov_salt);
+}
+
+int bt_mesh_prov_conf_salt(const uint8_t conf_inputs[145], uint8_t salt[16])
+{
+	const uint8_t conf_salt_key[16] = { 0 };
+
+	return bt_mesh_aes_cmac_one(conf_salt_key, conf_inputs, 145, salt);
+}
+
+int bt_mesh_prov_conf_key(const uint8_t dhkey[32], const uint8_t conf_salt[16],
+			  uint8_t conf_key[16])
+{
+	return bt_mesh_k1(dhkey, 32, conf_salt, "prck", conf_key);
+}
+
+int bt_mesh_prov_conf(const uint8_t conf_key[16], const uint8_t rand[16],
+		      const uint8_t auth[16], uint8_t conf[16])
+{
+	struct bt_mesh_sg sg[] = { { rand, 16 }, { auth, 16 } };
+
+	BT_DBG("ConfirmationKey %s", bt_hex(conf_key, 16));
+	BT_DBG("RandomDevice %s", bt_hex(rand, 16));
+	BT_DBG("AuthValue %s", bt_hex(auth, 16));
+
+	return bt_mesh_aes_cmac(conf_key, sg, ARRAY_SIZE(sg), conf);
+}
+
+int bt_mesh_prov_decrypt(const uint8_t key[16], uint8_t nonce[13],
+			 const uint8_t data[25 + 8], uint8_t out[25])
+{
+	return bt_ccm_decrypt(key, nonce, data, 25, NULL, 0, out, 8);
```

Comment:
This is for both decrypt/encrypt, is the data size of 25 and out size of 8 correct when the arrays are 33 and 25?

Reply:
yep. this is authentication calculation. 8 bytes is mic in Provisioning Confirmation frame.